Replace individual `reporter` keys with a document top-level `owners` key

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request replaces tracking the `reporter` for each product entry with tracking it on a per-file basis using an `owners` key that is a list of data owners. Changes are made to support using this new key to populate the `Reporter` field when outputting to Markdown as well as collating an de-duplicating entries when merging YAML files.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This makes it clear that it is not intended for submitters to modify who is reporting data. That field is meant to inform end consumers who own the data so change requests can be taken back to the source.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass. I tested this functionality in [cisagov/github-actions-playerground/testing/log4j_list_update](https://github.com/cisagov/github-actions-playground/tree/testing/log4j_list_update) where you can see the resultant files. I tested the de-duplication and owner collating logic locally to handle cases not seen with the setup in the aforementioned branch.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

mcdonnnj committed:
Use a top level key to store data ownership

Instead of creating a `reporter` key for every entry in the converted
Markdown we instead establish data ownership at in the YAML file as a
whole. This will ensure that all data in a given YAML file is attributed
to a specific source as was intended by the `reporter` key.

mcdonnnj committed:
Update to handle new `owners` key in YAML format

This updates the normalize-yml and yml2md scripts to work with the
addition of the owners key to the YAML format. This includes handling
different owners per ingested file, multiple owners in an ingested
file, and de-duplicating owner entries for output.

mcdonnnj committed:
Update the README to remove `reporter` key

Remove the `reporter` key from the `software` example entry in the
README and make sure the keys are correctly sorted.

dav3r commented on src/yml/normalize_yml.py:
```diff
@@ -36,13 +36,20 @@
+def munge(
+    filenames: list[str], owners: list[dict[str, str]] = [], canonical=False
+) -> Software:
     """Munge together the "software" nodes from YAML files into a single Python dictionary."""
```
Update docstring to reflect additional functionality:
```suggestion
    """Munge together the "owners" and "software" nodes from YAML files into a single Python dictionary."""
```

dav3r reviewed (approved):
Looks good, makes sense.  👍

mcdonnnj committed:
Update docstring to reflect new functionality

Update the docstring for the normalize_yml.munge function to reflect
the functionality added related to the document top-level "owners" key.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k commented on src/yml/normalize_yml.py:
```diff
@@ -36,13 +36,20 @@
+def munge(
+    filenames: list[str], owners: list[dict[str, str]] = [], canonical=False
+) -> Software:
+    """Munge together the "owners" and "software" nodes from YAML files into a single Python dictionary."""
```
I don't think that "from YAML files" applies to the `owners` information, but the current comment makes it sound like that.  I suggest increasing clarity with something like this: 
```suggestion
    """Combine into a single Python dictionary the owner and software information, reading in the latter from the specified YAML files."""
```

mcdonnnj replied:
The `owners` information is pulled in from a top-level key in the YAML file(s). Any `owners` keys in processed YAML files are collated and de-duplicated into one list.

jsf9k replied:
Ah, OK.

jsf9k reviewed (changes requested):
One minor request to tweak a comment, otherwise this looks grand to me.

mcdonnnj committed:
Refactor normalize-yml script

Refactor the logic in the normalize-yml script to treat the "owners"
key data consistently. This is a result of lgtm warning about modifying
a parameter with a default value. Relying on Python's pass-by-reference
is also a somewhat ugly way to approach handling that data in the first
place.

dav3r reviewed (approved):
Nice refactoring to clean this up. 👍
